### ISSUE TITLE
urInit with no adapters

### ISSUE THREAD

**Issue by smaslov-intel** [oneapi-src/unified-runtime#518]
Currently `urInit` would return UR_ERROR_UNINITIALIZED when no adapters were discovered. I suggest that we either 1) return a more descriptive error e.g. UR_ERROR_NO_ADAPTERS, or 2) return success, and then return 0 platforms in succeeding calls to `urPlatformGet`, if any.

**Comment by jandres742** [comment]
@smaslov-intel : if there are no adapters found at urInit time, is there any point on continuing? 
Could that new error, NO_ADAPTERS, be replaced by a descriptive message from urGetLastResult?

**Comment by smaslov-intel** [comment]
> Could that new error, NO_ADAPTERS, be replaced by a descriptive message from urGetLastResult?

Yes, that's a possibility. Still, UR L0 should set an explanatory error, and return UR_RESULT_ERROR_ADAPTER_SPECIFIC then

**Comment by pbalcer** [comment]
Hmm, the second option could be more straightforward. The language runtime wouldn't have to special case the error and would just handle getting 0 platforms, which unambiguously identifies the problem.

**Comment by alycm** [comment]
Discussion call 2023-05-17:

- Why is it an error for there to be zero adapters?  Query adapters, get none, but it's not an error.
- But why?  What would you be able to do, why let the application get that far?
- The loader is what returns `UR_ERROR_UNINITIALIZED`.

Most agree that reporting 0 adapters is valid.
